File: data_structures/map/applied/map_index/four_sum.scala
object Solution {
    def fourSum(nums: Array[Int], target: Int): List[List[Int]] = {
        // greedy or dynamic programming? probably latter because can do two-sum approach ++
        
        // two sum has 3 approaches
        // 1. brute force 2x for loops
        // 2. hashmap
        // 3. sort + two-pointers
        
        // four-sum
        // 1. 2 for loops + hashmap - O(n^3)
        // 2. 2 for loops + two-pointers - ~O(n^3)
        // 3. other approaches? build a map of (sum -> (i, j)) to indices O(n^2) - then iterate through keys (i.e. sums) to see if (target - sum exists)
        
        import scala.collection.mutable.HashMap
        import scala.collection.mutable.Set
        var pairMap = HashMap[Int, List[(Int, Int)]]()
        for(i <- 0 until nums.size){
            for(j <- i + 1 until nums.size){
                // problem is if have two sums the same we only keep one and we need to find all the solutions
                val pairSum = nums(i) + nums(j)
                pairMap = pairMap += (pairSum -> ((i,j) :: pairMap.getOrElse(pairSum,List[(Int, Int)]())))
            }
        }
        pairMap.keySet
             .foldLeft(Set[List[Int]]()){
                 (solutions, pairSum) => 
                    pairMap.get(target - pairSum)
                            // need to check for duplicate indices
                           .map{listComplementIndices => listComplementIndices
                                .foreach{ case (complementIndex1: Int, complementIndex2: Int) => 
                                    val listIndices = pairMap.getOrElse(pairSum, List[(Int, Int)]())
                                    listIndices.foreach {
                                        case (index1, index2) =>
                                            if(List(index1, index2, complementIndex1, complementIndex2).distinct.length == 4)
                                                solutions.addOne(List(nums(index1), nums(index2), nums(complementIndex1), nums(complementIndex2)).sorted) 
                                    }
                                }
                                solutions
                                    
                            }.getOrElse(solutions)

             }.toList
    }
}

File: data_structures/sorted/applied/two_sum_two_pointers/four_sum.go
import "sort"

func fourSum(nums []int, target int) [][]int {
	// 1. 2 for loops + 2-sum hashmap O(n^3)
	// 2. 2 for loops + 2-sum 2-pointers O(n^3)
	n := len(nums)
	if nums == nil || n < 4 {
		return nil
	}
	sort.Ints(nums)
	var solution [][]int
	// two for loops
	for i := 0; i < n-3; i++ {
		for j := i + 1; j < n-2; j++ {
			// 2-pointer
			remainingTarget := target - nums[i] - nums[j]
			left, right := j+1, n-1
			for left < right {
				sum := nums[left] + nums[right]
				if sum < remainingTarget {
					left++
				} else if sum > remainingTarget {
					right--
				} else {
					quadruplet := []int{nums[i], nums[j], nums[left], nums[right]}
					solution = append(solution, quadruplet)
					// skip through duplicates
					for left < right && nums[left] == quadruplet[2] {
						left++
					}
					for left < right && nums[right] == quadruplet[3] {
						right--
					}
				}
			}
			// skip through duplicates
			for j < n-1 && nums[j] == nums[j+1] {
				j++
			}
		}
		// skip through duplicates
		for i < n-1 && nums[i] == nums[i+1] {
			i++
		}
	}
	return solution
}